stl/src/multprec.cpp

```diff
@@ -1,190 +1,63 @@
-    while (u[ulen - 1] == 0) {
-        --ulen;
-    }
-
-    return ulen;
+// TRANSITION, ABI: preserved for binary compatibility
+_CRTIMP2_PURE void __CLRCALL_PURE_OR_CDECL _MP_Mul(_MP_arr _Wx, const uint64_t _Prev, const uint64_t _Ax) noexcept {
+    // set _Wx to the 128-bit product of _Prev and _Ax
+    const auto _Result = _Unsigned128{_Prev} * _Ax;
+    assign_mp_from_u128(_Wx, _Result);
 }
 
 // TRANSITION, ABI: preserved for binary compatibility
-_CRTIMP2_PURE void __CLRCALL_PURE_OR_CDECL _MP_Rem(
-    _MP_arr u, unsigned long long v0) noexcept { // divide multi-word value by value, leaving remainder in u
-    unsigned long long v[2];
-    v[0]        = v0 & mask;
-    v[1]        = v0 >> shift;
-    const int n = limit(v, 2);
-    _Analysis_assume_(n > 0);
-    _Analysis_assume_(n <= 2);
-    const int m = limit(u, _MP_len) - n;
-    _Analysis_assume_(m > 0);
-    _Analysis_assume_(m <= _MP_len - n);
-
-    // Knuth, vol. 2, p. 272, Algorithm D
-    // D1: [Normalize.]
-    unsigned long long d = maxVal / (v[n - 1] + 1);
-    if (d != 1) { // scale numerator and divisor
-        mul(u, _MP_len, d);
-        mul(v, n, d);
-    }
-    // D2: [Initialize j.]
-    for (int j = m; 0 <= j; --j) { // D3: [Calculate qh.]
-        unsigned long long qh = ((u[j + n] << shift) + u[j + n - 1]) / v[n - 1];
-        if (qh == 0) {
-            continue;
-        }
-
-        unsigned long long rh = ((u[j + n] << shift) + u[j + n - 1]) % v[n - 1];
-        for (;;) {
-#pragma warning(push)
-#pragma warning(disable : 6385) // TRANSITION, GH-1008
-            if (qh < maxVal && qh * v[n - 2] <= (rh << shift) + u[j + n - 2]) {
-#pragma warning(pop)
-                break;
-            } else { // reduce tentative value and retry
-                --qh;
-                rh += v[n - 1];
-                if (maxVal <= rh) {
-                    break;
-                }
-            }
-        }
-
-        // D4: [Multiply and subtract.]
-        unsigned long long k = 0;
-        int i;
-        for (i = 0; i < n; ++i) { // multiply and subtract
-            u[j + i] -= qh * v[i] + k;
-            k = u[j + i] >> shift;
-            if (k) {
-                k = maxVal - k;
-            }
-
-            u[j + i] &= mask;
-        }
-        for (; k != 0 && j + i < _MP_len; ++i) { // propagate borrow
-            u[j + i] -= k;
-            k = u[j + i] >> shift;
-            if (k) {
-                k = maxVal - k;
-            }
-
-            u[j + i] &= mask;
-        }
-        // D5: [Test remainder.]
-        if (k != 0) { // D6: [Add back.]
-            --qh;
-            add(u + j, n + 1, v, n);
-        }
-        // D7: [Loop on j.]
-    }
-    // D8: [Unnormalize.]
-    if (d != 1) {
-        div(u, d);
-    }
+_CRTIMP2_PURE void __CLRCALL_PURE_OR_CDECL _MP_Rem(_MP_arr _Wx, const uint64_t _Mx) noexcept {
+    // perform "_Wx %= _Mx"
+    const auto _Result = u128_from_mp(_Wx) % _Mx;
+    assign_mp_from_u128(_Wx, _Result);

```

Comment:
minor inconsistency: _Ugly_variables, non_ugly_functions

Reply:
Good point, that's a relic from when I was using pretty parameters, before I changed them to align with header usage.